File: src/open_workshop_storage/core/blurhash_cache.py
from __future__ import annotations

import hashlib
import json
from typing import Any, Optional


def build_blurhash_cache_key(
    real_path: str,
    mtime_ns: int,
    size: int,
    components_x: int = 4,
    components_y: int = 3,
    max_dimension: int = 64,
) -> str:
    payload = {
        "path": real_path,
        "mtime_ns": int(mtime_ns),
        "size": int(size),
        "components_x": int(components_x),
        "components_y": int(components_y),
        "max_dimension": int(max_dimension),
    }
    raw = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()


def encode_blurhash_cache_value(blurhash: str, width: int, height: int) -> dict[str, Any]:
    return {
        "blurhash": blurhash,
        "width": int(width),
        "height": int(height),
    }


def decode_blurhash_cache_value(data: dict[str, Any]) -> Optional[tuple[str, int, int]]:
    try:
        blurhash = str(data["blurhash"])
        width = int(data["width"])
        height = int(data["height"])
    except (KeyError, TypeError, ValueError):
        return None
    if not blurhash:
        return None
    return blurhash, width, height

File: src/blurhash_cache.rs
use serde_json::{Map, Value};
use sha2::{Digest, Sha256};

pub fn build_blurhash_cache_key(
    real_path: &str,
    mtime_ns: u128,
    size: u64,
    components_x: u32,
    components_y: u32,
    max_dimension: u32,
) -> String {
    let payload = serde_json::json!({
        "path": real_path,
        "mtime_ns": mtime_ns,
        "size": size,
        "components_x": components_x,
        "components_y": components_y,
        "max_dimension": max_dimension,
    });
    let raw = serde_json::to_string(&payload).unwrap_or_default();
    let digest = Sha256::digest(raw.as_bytes());
    format!("{digest:x}")
}

pub fn encode_blurhash_cache_value(blurhash: &str, width: u32, height: u32) -> Map<String, Value> {
    let mut map = Map::new();
    map.insert("blurhash".to_string(), Value::String(blurhash.to_string()));
    map.insert("width".to_string(), Value::from(width));
    map.insert("height".to_string(), Value::from(height));
    map
}

pub fn decode_blurhash_cache_value(data: &Map<String, Value>) -> Option<(String, u32, u32)> {
    let blurhash = data.get("blurhash")?.as_str()?.to_string();
    if blurhash.is_empty() {
        return None;
    }
    let width = data.get("width")?.as_u64()? as u32;
    let height = data.get("height")?.as_u64()? as u32;
    Some((blurhash, width, height))
}
